src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java

```diff
@@ -0,0 +1,105 @@
+package ch.uzh.ifi.hase.soprafs22.game.logger.implementation;
+
+import ch.uzh.ifi.hase.soprafs22.game.logger.GameLogger;
+import ch.uzh.ifi.hase.soprafs22.game.logger.GameStatistics;
+
+import java.util.*;
+
+/**
+ * Implementation of GameLogger to provide GameStatistics.
+ */
+public final class GameLoggerImpl implements GameStatistics, GameLogger {
+    private final Map<Long, Integer> unitsPerPlayer;
+
+    private final List<Map<Long, Integer>> turnSnapshots = new ArrayList<>();
+    private int totalMoves = 0;
+    private int turn = 0;
+
+    public static Builder newBuilder(){
+        return new Builder();
+    }
+
+    public static final class Builder {
```

Comment:
also here, I am not familiar with a nested Builder.
Why don't we use the simple approach with a public constructor of the outer class?

Reply:
I just copied the structure of AutoValue.Builder (see https://github.com/google/auto/blob/master/value/userguide/builders.md). Nesting the Builder inside the class allows keeping the constructor private, so the only way to get those objects is through the builder.

I deleted the builder, they are only to make code easier to read and this one seems to be a problem.